examples/keras_io/tensorflow/nlp/neural_machine_translation_with_transformer.py

```diff
@@ -0,0 +1,477 @@
+train_spa_texts = [pair[1] for pair in train_pairs]
+eng_vectorization.adapt(train_eng_texts)
+spa_vectorization.adapt(train_spa_texts)
+
+"""
+Next, we'll format our datasets.
+
+At each training step, the model will seek to predict target words N+1 (and beyond)
+using the source sentence and the target words 0 to N.
+
+As such, the training dataset will yield a tuple `(inputs, targets)`, where:
+
+- `inputs` is a dictionary with the keys `encoder_inputs` and `decoder_inputs`.
+`encoder_inputs` is the vectorized source sentence and `encoder_inputs` is the target sentence "so far",
+that is to say, the words 0 to N used to predict word N+1 (and beyond) in the target sentence.
+- `target` is the target sentence offset by one step:
+it provides the next words in the target sentence -- what the model will try to predict.
+"""
+
+import tensorflow.data as tf_data
+
+def format_dataset(eng, spa):
+    eng = eng_vectorization(eng)
+    spa = spa_vectorization(spa)
+    return (
+        {
+            "encoder_inputs": eng,
+            "decoder_inputs": spa[:, :-1],
+        },
+        spa[:, 1:],
+    )
+
+
+def make_dataset(pairs):
+    eng_texts, spa_texts = zip(*pairs)
+    eng_texts = list(eng_texts)
+    spa_texts = list(spa_texts)
+    dataset = tf_data.Dataset.from_tensor_slices((eng_texts, spa_texts))
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(format_dataset)
+    return dataset.shuffle(2048).prefetch(16).cache()
+
+
+train_ds = make_dataset(train_pairs)
+val_ds = make_dataset(val_pairs)
+
+"""
+Let's take a quick look at the sequence shapes
+(we have batches of 64 pairs, and all sequences are 20 steps long):
+"""
+
+for inputs, targets in train_ds.take(1):
+    print(f'inputs["encoder_inputs"].shape: {inputs["encoder_inputs"].shape}')
+    print(f'inputs["decoder_inputs"].shape: {inputs["decoder_inputs"].shape}')
+    print(f"targets.shape: {targets.shape}")
+
+"""
+## Building the model
+
+Our sequence-to-sequence Transformer consists of a `TransformerEncoder`
+and a `TransformerDecoder` chained together. To make the model aware of word order,
+we also use a `PositionalEmbedding` layer.
+
+The source sequence will be pass to the `TransformerEncoder`,
+which will produce a new representation of it.
+This new representation will then be passed
+to the `TransformerDecoder`, together with the target sequence so far (target words 0 to N).
+The `TransformerDecoder` will then seek to predict the next words in the target sequence (N+1 and beyond).
+
+A key detail that makes this possible is causal masking
+(see method `get_causal_attention_mask()` on the `TransformerDecoder`).
+The `TransformerDecoder` sees the entire sequences at once, and thus we must make
+sure that it only uses information from target tokens 0 to N when predicting token N+1
+(otherwise, it could use information from the future, which would
+result in a model that cannot be used at inference time).
+"""
+import keras_core.ops as ops
```

Comment:
`from keras_core import ops` + move import to top

Reply:
done